import { UserRepository } from "@/core/domain/ports/UserRepository";
import { Error } from "@/core/domain/models/Error";
import { Response } from "@/core/domain/models/Response";

export class AssignGroupToUser {
    constructor(private userRepository: UserRepository) {}

    async execute(userId: string, groupId: string[]): Promise<Response> {
        try {
            const response = await this.userRepository.assignUserGroup(userId, groupId);
            
            return {
                status: response.status,
                path: "/",
                message: "GROUP_ASSIGNED",
            };
        } catch (err) {
            const error = err as Error;
            if (error.type === "api") {
                if(error.status === 403) {
                    throw {
                        status: "ACCESS_DENIED",
                        message: error.message,
                    };
                } else if (error.status === 400) {
                    throw {
                        status: "VALIDATION_ERROR",
                        message: error.message,
                    };
                } else if (error.status === 429) {
                    throw {
                        status: "TOO_MANY_REQUESTS",
                        message: error.message,
                    };
                } else if (error.status === 401) {
                    throw {
                        status: "UNAUTHORIZED",
                        message: error.message,
                    };
                } else if (error.status === 404) {
                    throw {
                        status: "NOT_FOUND",
                        message: error.message,
                    };
                }
            }
            if (error.type === "validation") {
                throw {
                    status: "VALIDATION_ERROR",
                    message: "La estructura de datos recibida no es válida.",
                };
            }
            if (error.type === "unknown_api_error") {
                throw {
                    status: "UNKNOWN_API_ERROR",
                    message: "La estructura de error de la API no es válida.",
                };
            }
            throw {
                status: "UNKNOWN_ERROR",
                message: "Error de red",
            };
        }
    }
}